File: androidApp/src/main/kotlin/it/mensa/app/features/profile/sub/PaymentMethodPickerSheet.kt
package it.mensa.app.features.profile.sub

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.AddCircle
import androidx.compose.material.icons.outlined.CreditCard
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.stripe.android.paymentsheet.rememberPaymentSheet
import it.mensa.app.support.tr
import org.koin.androidx.compose.koinViewModel

/**
 * Inline payment-method picker. Mirrors Flutter's `PaymentMethodPicker`:
 *
 *   - On open, shows the default method + a "Pay {amount}" CTA. The
 *     "Cambia metodo" link toggles to a radio list of saved methods.
 *   - "Aggiungi metodo" launches Stripe's PaymentSheet in setup-intent
 *     mode and reloads on success.
 *   - The caller passes `onConfirm` to receive the selected method id —
 *     the donation screen then triggers the actual donate flow.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PaymentMethodPickerSheet(
    amountLabel: String,
    onDismiss: () -> Unit,
    onConfirm: (paymentMethodId: String) -> Unit,
    vm: PaymentMethodPickerViewModel = koinViewModel(),
) {
    val uiState by vm.uiState.collectAsStateWithLifecycle()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val colorScheme = MaterialTheme.colorScheme

    // SetupIntent → PaymentSheet for adding a new method. Compose hosts
    // the launcher and pipes results back into the VM.
    val paymentSheet = rememberPaymentSheet(vm::onAddMethodResult)
    LaunchedEffect(uiState.addPending) {
        val pending = uiState.addPending ?: return@LaunchedEffect
        paymentSheet.presentWithSetupIntent(
            setupIntentClientSecret = pending.clientSecret,
            configuration = pending.configuration,
        )
        vm.consumeAddPending()
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 20.dp, vertical = 8.dp),
        ) {
            Text(
                tr("app.payments.title", fallback = "Metodo di pagamento"),
                style = MaterialTheme.typography.titleLarge,
                fontWeight = FontWeight.SemiBold,
            )

            Spacer(Modifier.height(16.dp))

            if (uiState.loading) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 32.dp),
                    contentAlignment = Alignment.Center,
                ) { CircularProgressIndicator() }
            } else {
                val default = vm.defaultMethod()

                if (default == null || uiState.showPicker) {
                    // Flutter shows the add-CTA whenever there's no default
                    // OR when the user is browsing the picker — same here.
                    AddMethodTile(
                        enabled = !uiState.adding,
                        onClick = vm::addMethod,
                    )
                    Spacer(Modifier.height(8.dp))
                }

                if (default != null && !uiState.showPicker) {
                    PaymentMethodRow(
                        method = default,
                        selected = true,
                        showRadio = false,
                        onClick = {},
                    )
                }

                if (default != null && uiState.showPicker) {
                    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                        uiState.methods.forEach { m ->
                            PaymentMethodRow(
                                method = m,
                                selected = m.id == uiState.defaultId,
                                showRadio = true,
                                onClick = { vm.selectMethod(m.id) },
                            )
                        }
                    }
                }

                Spacer(Modifier.height(16.dp))

                if (!uiState.showPicker) {
                    Button(
                        onClick = {
                            default?.let { onConfirm(it.id) }
                        },
                        enabled = default != null,
                        modifier = Modifier
                            .fillMaxWidth()
                            .defaultMinSize(minHeight = 56.dp),
                    ) {
                        Text(
                            text = tr(
                                "app.payments.pay_amount",
                                fallback = "Paga {amount}",
                                "amount" to amountLabel,
                            ),
                        )
                    }
                    if (uiState.methods.size > 1 || default != null) {
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(top = 4.dp),
                            horizontalArrangement = Arrangement.End,
                        ) {
                            TextButton(onClick = vm::toggleShowPicker) {
                                Text(
                                    tr(
                                        "app.payments.change_method",
                                        fallback = "Cambia metodo di pagamento",
                                    ),
                                    style = MaterialTheme.typography.labelMedium,
                                )
                            }
                        }
                    }
                } else {
                    TextButton(
                        onClick = vm::toggleShowPicker,
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        Text(tr("app.common.back", fallback = "Indietro"))
                    }
                }
            }

            Spacer(Modifier.height(8.dp))
        }
    }
}

@Composable
private fun AddMethodTile(
    enabled: Boolean,
    onClick: () -> Unit,
) {
    val colorScheme = MaterialTheme.colorScheme
    Card(
        onClick = onClick,
        enabled = enabled,
        colors = CardDefaults.cardColors(
            containerColor = colorScheme.surfaceContainerLow,
        ),
        modifier = Modifier.fillMaxWidth(),
    ) {
        Row(
            modifier = Modifier.padding(16.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                tr("app.payments.add", fallback = "Aggiungi metodo di pagamento"),
                style = MaterialTheme.typography.bodyLarge,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier.weight(1f),
            )
            Icon(
                Icons.Outlined.AddCircle,
                contentDescription = null,
                tint = colorScheme.primary,
            )
        }
    }
}

@Composable
private fun PaymentMethodRow(
    method: it.mensa.shared.model.PaymentMethodModel,
    selected: Boolean,
    showRadio: Boolean,
    onClick: () -> Unit,
) {
    val colorScheme = MaterialTheme.colorScheme
    Card(
        onClick = onClick,
        colors = CardDefaults.cardColors(
            containerColor = if (selected) colorScheme.secondaryContainer
                             else colorScheme.surfaceContainerLow,
        ),
        modifier = Modifier.fillMaxWidth(),
    ) {
        Row(
            modifier = Modifier.padding(14.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Surface(
                shape = CircleShape,
                color = colorScheme.primaryContainer,
                modifier = Modifier.size(36.dp),
            ) {
                Box(contentAlignment = Alignment.Center) {
                    Icon(
                        Icons.Outlined.CreditCard,
                        contentDescription = null,
                        tint = colorScheme.onPrimaryContainer,
                        modifier = Modifier.size(18.dp),
                    )
                }
            }
            Spacer(Modifier.size(12.dp))
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    method.brand.ifEmpty { "Carta" }.replaceFirstChar { it.uppercase() },
                    style = MaterialTheme.typography.titleSmall,
                )
                Text(
                    method.display.ifEmpty { "•••• ••••" },
                    style = MaterialTheme.typography.bodySmall,
                    color = colorScheme.onSurfaceVariant,
                )
            }
            if (showRadio) {
                RadioButton(selected = selected, onClick = onClick)
            }
        }
    }
}

File: iosApp/iosApp/Features/Profile/Sub/PaymentMethodPickerSheet.swift
import SwiftUI
import Shared
import StripePaymentSheet

/// Inline payment-method picker. Mirrors Flutter's `PaymentMethodPicker`:
///
///   - On open, shows the default method + a "Paga {amount}" CTA.
///   - "Cambia metodo" toggles to a radio list of saved methods —
///     selecting one calls `/api/payment/default`.
///   - "Aggiungi metodo" runs Stripe's PaymentSheet in setup-intent mode
///     and reloads on success.
///   - On confirm, the caller receives the chosen payment method id and
///     triggers the donation (PaymentIntent + confirmPayment).
struct PaymentMethodPickerSheet: View {
    let amountLabel: String
    let onConfirm: (_ paymentMethodId: String) -> Void
    let onDismiss: () -> Void

    @State private var vm = PaymentMethodPickerViewModel()

    var body: some View {
        NavigationStack {
            content
                .navigationTitle(tr("app.payments.title",
                                    fallback: "Metodo di pagamento"))
                .navigationBarTitleDisplayMode(.inline)
                .toolbar {
                    ToolbarItem(placement: .topBarTrailing) {
                        Button(action: onDismiss) {
                            Image(systemName: "xmark.circle.fill")
                                .foregroundStyle(.secondary)
                        }
                    }
                }
        }
        .task { await vm.load() }
        .alert("Errore",
               isPresented: Binding(
                get: { vm.errorMessage != nil },
                set: { if !$0 { vm.errorMessage = nil } }
               )) {
            Button("OK", role: .cancel) { vm.errorMessage = nil }
        } message: {
            Text(vm.errorMessage ?? "")
        }
    }

    @ViewBuilder
    private var content: some View {
        ScrollView {
            VStack(spacing: 12) {
                if vm.loading {
                    ProgressView()
                        .padding(.vertical, 40)
                } else {
                    let defaultMethod = vm.defaultMethod()

                    if defaultMethod == nil || vm.showPicker {
                        addMethodTile
                    }

                    if let m = defaultMethod, !vm.showPicker {
                        PaymentMethodRow(method: m, selected: true, showRadio: false)
                            .padding(.horizontal, 4)
                    }

                    if defaultMethod != nil && vm.showPicker {
                        VStack(spacing: 6) {
                            ForEach(vm.methods, id: \.id) { m in
                                PaymentMethodRow(
                                    method: m,
                                    selected: m.id == vm.defaultId,
                                    showRadio: true
                                )
                                .padding(.horizontal, 4)
                                .contentShape(Rectangle())
                                .onTapGesture { Task { await vm.selectMethod(id: m.id) } }
                            }
                        }
                    }

                    if !vm.showPicker {
                        Button {
                            if let m = defaultMethod { onConfirm(m.id) }
                        } label: {
                            Text(tr("app.payments.pay_amount",
                                    fallback: "Paga {amount}",
                                    ["amount": amountLabel]))
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 14)
                                .background(AppTheme.brandGradient,
                                            in: RoundedRectangle(cornerRadius: 14))
                                .foregroundStyle(.white)
                                .fontWeight(.semibold)
                        }
                        .disabled(defaultMethod == nil)
                        .opacity(defaultMethod == nil ? 0.5 : 1.0)
                        .padding(.top, 8)

                        if !vm.methods.isEmpty {
                            Button(action: { vm.showPicker.toggle() }) {
                                Text(tr("app.payments.change_method",
                                        fallback: "Cambia metodo di pagamento"))
                                    .font(.footnote)
                            }
                            .padding(.top, 2)
                        }
                    } else {
                        Button(action: { vm.showPicker.toggle() }) {
                            Text(tr("app.common.back", fallback: "Indietro"))
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 12)
                        }
                        .buttonStyle(.bordered)
                        .padding(.top, 8)
                    }
                }
            }
            .padding(20)
        }
    }

    private var addMethodTile: some View {
        Button {
            vm.addMethod()
        } label: {
            HStack {
                Text(tr("app.payments.add",
                        fallback: "Aggiungi metodo di pagamento"))
                    .fontWeight(.semibold)
                Spacer()
                Image(systemName: "plus.circle.fill")
                    .foregroundStyle(AppTheme.Colors.brandTintAdaptive)
            }
            .padding(16)
            .background(Color(.secondarySystemBackground),
                        in: RoundedRectangle(cornerRadius: 12))
            .foregroundStyle(.primary)
        }
        .disabled(vm.adding)
    }
}

private struct PaymentMethodRow: View {
    let method: PaymentMethodModel
    let selected: Bool
    let showRadio: Bool

    var body: some View {
        HStack(spacing: 12) {
            ZStack {
                Circle()
                    .fill(Color.accentColor.opacity(0.15))
                Image(systemName: "creditcard.fill")
                    .foregroundStyle(.tint)
            }
            .frame(width: 36, height: 36)

            VStack(alignment: .leading, spacing: 2) {
                Text(method.brand.isEmpty
                     ? "Carta"
                     : method.brand.prefix(1).uppercased() + method.brand.dropFirst())
                    .font(.subheadline.weight(.semibold))
                Text(method.display.isEmpty ? "•••• ••••" : method.display)
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

            Spacer()

            if showRadio {
                Image(systemName: selected ? "largecircle.fill.circle" : "circle")
                    .foregroundStyle(selected ? Color.accentColor : Color.secondary)
            }
        }
        .padding(14)
        .background(
            (selected
             ? Color.accentColor.opacity(0.1)
             : Color(.secondarySystemBackground)),
            in: RoundedRectangle(cornerRadius: 12)
        )
    }
}

// MARK: - View model

@MainActor
@Observable
final class PaymentMethodPickerViewModel {
    var loading = true
    var methods: [PaymentMethodModel] = []
    var defaultId: String?
    var showPicker = false
    var adding = false
    var errorMessage: String?

    func load() async {
        loading = true
        defer { loading = false }
        async let methodsTask: [PaymentMethodModel]? = try? await koin.paymentMethods.refresh()
        async let customerTask: StripeCustomerObject? = try? await koin.paymentMethods.customer()
        let m = await methodsTask ?? []
        let c = await customerTask
        methods = m
        defaultId = c?.defaultPaymentMethodId
    }

    func defaultMethod() -> PaymentMethodModel? {
        guard let id = defaultId else { return nil }
        return methods.first { $0.id == id }
    }

    func selectMethod(id: String) async {
        loading = true
        defer { loading = false }
        do {
            try await koin.paymentMethods.setDefault(paymentMethodId: id)
            defaultId = id
            showPicker = false
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func addMethod() {
        if adding { return }
        adding = true
        StripeService.addPaymentMethod { [weak self] result in
            guard let self else { return }
            adding = false
            switch result {
            case .completed:
                Task { await self.load() }
            case .canceled:
                break
            case .failed(let error):
                errorMessage = error.localizedDescription
            }
        }
    }
}
